pkg/restore/restore.go

```diff
@@ -917,43 +939,59 @@ func (r *pvRestorer) executePVAction(obj *unstructured.Unstructured) (*unstructu
 		return obj, nil
 	}
 
-	// If we can't find a snapshot record for this particular PV, it most likely wasn't a PV that Ark
-	// could snapshot, so return early instead of trying to restore from a snapshot.
-	backupInfo, found := r.volumeBackups[pvName]
-	if !found {
+	log := r.logger.WithFields(logrus.Fields{"persistentVolume": pvName})
+
+	var foundSnapshot *volume.Snapshot
+	for _, snapshot := range r.volumeSnapshots {
+		if snapshot.Spec.PersistentVolumeName == pvName {
+			foundSnapshot = snapshot
+			break
+		}
+	}
+	if foundSnapshot == nil {
+		log.Info("skipping no snapshot found")
 		return obj, nil
 	}
 
-	// Past this point, we expect to be doing a restore
+	location, err := r.snapshotLocationLister.VolumeSnapshotLocations(r.backupNamespace).Get(foundSnapshot.Spec.Location)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
 
-	if r.blockStore == nil {
-		return nil, errors.New("you must configure a persistentVolumeProvider to restore PersistentVolumes from snapshots")
+	blockStore, err := r.blockStoreGetter.GetBlockStore(location.Spec.Provider)
+	if err != nil {
+		return nil, errors.WithStack(err)
 	}
 
-	log := r.logger.WithFields(
-		logrus.Fields{
-			"persistentVolume": pvName,
-			"snapshot":         backupInfo.SnapshotID,
-		},
-	)
+	if err := blockStore.Init(location.Spec.Config); err != nil {
+		return nil, errors.WithStack(err)
+	}
 
-	log.Info("restoring persistent volume from snapshot")
-	volumeID, err := r.blockStore.CreateVolumeFromSnapshot(backupInfo.SnapshotID, backupInfo.Type, backupInfo.AvailabilityZone, backupInfo.Iops)
+	volumeID, err := blockStore.CreateVolumeFromSnapshot(foundSnapshot.Status.ProviderSnapshotID, foundSnapshot.Spec.VolumeType, foundSnapshot.Spec.VolumeAZ, foundSnapshot.Spec.VolumeIOPS)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithStack(err)
 	}
+
+	log = log.WithFields(logrus.Fields{"snapshot": foundSnapshot.Status.ProviderSnapshotID})
 	log.Info("successfully restored persistent volume from snapshot")
+	// used to be update1 which is then cast to Unstructured and returned
+	updated1, err := blockStore.SetVolumeID(obj, volumeID)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
 
-	updated1, err := r.blockStore.SetVolumeID(obj, volumeID)
+	verifyVolumeID, err := blockStore.GetVolumeID(updated1)
```

Comment:
was there a specific reason to have this `GetVolumeID` call here?  I don't think it should be necessary, since the `SetVolumeID` call above will return an error if it can't set it properly, but I might be overlooking something

Reply:
That is just oversight from when I was debugging some test failures.